### DISCUSSION TITLE
Templating via CRS

### DISCUSSION THREAD

**Discussion by bigkevmcd** [kubernetes-sigs/cluster-api#4906] [Ideas]
I'm wondering if there'd be any interest in providing templating of the resources defined in a CRS.

For example...

```yaml
apiVersion: addons.cluster.x-k8s.io/v1alpha3
kind: ClusterResourceSet
metadata:
  name: example-crs
  namespace: default
spec:
  clusterSelector:
    matchLabels:
      bootstrap.containers: yes
  resources:
  - kind: ConfigMap
    name: bootstrap-containers
```

If the resources defined in the _bootstrap-containers_ `ConfigMap` had something like `${annotations.github-repository}` or `${annotations.repository-path}` then somewhere around the point where the [resources are being applied](https://github.com/kubernetes-sigs/cluster-api/blob/master/exp/addons/controllers/clusterresourceset_helpers.go#L57) it could do a replace on the objects.

The fields `annotations.X` and perhaps `labels.X` would lookup on the annotations and labels applied to the _Cluster_ object that the CRS applies to and insert those values into the cluster.

This would be useful if you wanted to bootstrap a workload into a newly-provisioned cluster, but customise the processes slightly per-cluster?

**Comment by fabriziopandini** [comment]
The current stance about addon management is summarized in https://github.com/kubernetes-sigs/cluster-api/blob/main/docs/proposals/20220712-cluster-api-addon-orchestration.md 

TL;DR;
- CRS are a very basic solution addon orchestration
- We should rely on tools existing on the ecosystem for more advanced use cases

https://github.com/kubernetes-sigs/cluster-api-addon-provider-helm is an example (but not the only one)
